asdf/extension/_compressor.py

```diff
@@ -0,0 +1,84 @@
+"""
+Compressor is an interface to implement extensions to the compression
+module. Extensions will typically subclass the Compressor ABC and
+provide that subclass as a setuptools entry point.  Decompressor
+is likewise used for decompression.
+
+Note that this interface has similar patterns to Converter.  This
+interface is designed for compression and decompression of ASDF
+binary array blocks, while Converter is designed for serialization
+of custom Python types into the YAML tree.
+"""
+
+import abc
+
+class Compressor(abc.ABC):
+    """
+    Abstract base class for plugins that compress binary data.
+
+    Implementing classes must provide the `labels` property, and
+    the `compress()` method.
+    May also provide a constructor that takes kwargs that are
+    set via `asdf.get_config().compression_options[label] = kwargs`.
+    """
+    @classmethod
+    def __subclasshook__(cls, C):
+        if cls is Compressor:
+            return (hasattr(C, "labels") and
+                    hasattr(C, "compress"))
+        return NotImplemented # pragma: no cover
+
+
+    @abc.abstractproperty
+    def labels(self):
+        """
+        Get the string labels that this Compressor
+        is able to compress.
+
+        Returns
+        -------
+        iterable of str
+            str labels handled by this class
+        """
+        pass # pragma: no cover
+
+
+    @abc.abstractmethod
+    def compress(self, data):
```

Comment:
I like the idea of configuring the compressors using`AsdfConfig`, but it would be nice if users in a console could interactively discover the available options without having to dig into the details of the `Compressor` subclass.  Maybe we should add a method here like `get_default_config` that returns the keys and default values (and maybe doc strings) for the compressor's config options.  Then we can use that method in AsdfConfig to generate structure that can be tab-completed in ipython.

Reply:
I like the tab-complete idea a lot.  Did you have an idea of how to register the tab-complete fields of a `Compressor` with `AsdfConfig`?  If `Compressor.get_default_config()` returns a key->(val,docstring) dict, then I guess when an `Extension` is loaded, it needs to register this dict with the config somehow.  Then the config can build an object with these fields and store it under `config.compression_options` such that
```
lzma_conf = config.compression_options['lzma']
lzma_conf.<TAB>
```
will prompt tab-completion.

Is `config.add_extension()` a good place to hook this logic?  Is it called for all extensions?

Reply:
Hmm... good question.  One option is to dynamically generate config classes for each compressor:

```python
label = "lzma"
config_dict = {
    "preset": (6, "LZMA compression preset.  Integer from 0-9."),
    "foo": ("bar", "Some other LZMA config parameter.")
}

cls = type(
    f"{label.upper()}CompressorConfig",
    (object,),
    {"__doc__": f"Configuration options for the {label.upper()} compressor."}
)

for name, (default_value, docstring) in config_dict.items():
    setattr(cls, name, property((lambda attr: lambda self: getattr(self, attr))(f"_{name}")))
    setattr(cls, name, getattr(cls, name).setter((lambda attr: lambda self, value: setattr(self, attr, value))(f"_{name}")))
    getattr(cls, name).__doc__ = docstring

def generate_init(config_dict):
    def init(self):
        for name, (default_value, docstring) in config_dict.items():
            setattr(self, f"_{name}", default_value)
    return init
cls.__init__ = generate_init(config_dict)
```
An instance of that class seems to behave like we'd want:
```python
In [82]: config = cls()

In [83]: config.
                 foo
                 preset
 
In [84]: ?config.preset
Type:        property
String form: <property object at 0x7ff4f60e2cc0>
Docstring:   LZMA compression preset.  Integer from 0-9.

In [85]: config.preset
Out[85]: 6

In [86]: config.preset = 9

In [87]: config.preset
Out[87]: 9
```

But now I'm wondering if instead we should ask the author of the compressor/extension provide an implementation of a config class.  That way they could e.g. perform validation in the setter, automatically adjust parameters with tied values, etc.  I can see that being a valuable general feature of `Extension`, not just for compression config options.

Reply:
I think I'm still wrestling with whether an extra layer of abstraction—a Config object, either automatically generated or supplied by the user—is likely to be helpful, or just extra boilerplate.  The tab-completion is nice, for sure, but this requires an extension to explicitly implement every arg supported by the underlying compression that an end user might want to use.  It's likely to be a lot of pass-through getters and setters, and copy-pasted docstrings from the underlying compressor.  It just seems verbose and fragile.

If we did go the dict-only approach, there's still some chance for validation in the `Compressor` constructor. You lose tab-completion and individual docstrings for the settings, but you also don't have to explicitly enumerate all settings and can point the user to the docs of the underlying compressor.

Reply:
Hmm... good point.  Trying to keep up with multiple versions of the same compression package with different options would make us crazy.  I do still like the idea of a general `Extension` feature that allows the extension to register a config class, but maybe in this case it would be something simple like this:

```python
class LzmaExtensionConfig:
  def __init__(self):
    self._compressor_options = {}
  
  @property
  def compressor_options(self):
    return self._compressor_options
    
  @compressor_options.setter
  def compressor_options(self, value):
     """
     Additional keyword arguments passed to `lzma.LZMACompressor`.  See
     https://docs.python.org/3/library/lzma.html for details.
     """
     self._compressor_options = value

class LzmaExtension:
  config_class = LzmaExtensionConfig
```

Then we add that to `AsdfConfig` somewhere reasonable so that users can do something like:

```python
asdf.get_config().extensions.<something>.compressor_options = {"preset": 9}
```

How does that strike you?  I don't know what `<something>` should be, the only reliably unique key to an extension is its URI, but that's a lot to type.

Reply:
I like that.  For the `<something>`, what about assigning a "short name" to each extension from the name of the entry point registration?  E.g. `lzma_ext` in this example:
```python
setup(
   ...
    entry_points={'asdf.extensions':['lzma_ext = mypkg.asdf:LzmaExtension']}
)
```
Short-name conflicts could be warned about and require a URI lookup.

But just to check, `asdf.get_config().extensions.lzma_ext` is a `LzmaExtension` object, not the `LzmaExtensionConfig`, right?  So I guess setting the config would require another level of indirection, like `asdf.get_config().extensions.lzma_ext.config`.  That's getting pretty verbose.  Maybe it could be shortened to `asdf.get_extension_config('lzma_ext')`?

Similarly, maybe the casual user won't care about which extension the compression came from, and should just be able to type: `asdf.get_compression_config('lzma')`.  I think it's just a question of API surface area versus verbosity.

Reply:
And formally, if a whole extension has one `compressor_config` associated with it, then we probably need to limit extensions to one compressor.  So packages that want to define multiple compressions need to define and register multiple extensions.  That's not the end of the world, just something worth thinking about.

Reply:
Oops, I forgot that `extensions` was already taken.  That hasn't been released yet, and was really only intended for use within the library.  We could change it to a `get_extensions` method instead to make access to the extension configs more convenient.

I was thinking the extension's config would be free-form.  It could have foo_compressor_config and bar_compressor_config properties if it wanted to, or any other name.  Then inside of the `FooCompressor` class the implementation would call `asdf.get_config().extensions.lzma.foo_compressor_config` to get the settings.

Reply:
@perrygreenfield has a use case for tweaking compression parameters on an individual array basis.  His example is lossy compression where the appropriate parameter values depend on the level of noise in a given array.  Given that, I wonder if our best option is to make the config an argument to the `AsdfFile.set_array_compression` method.  Something would also need to be done for the `all_array_compression` argument to `AsdfFile.write_to` and `AsdfFile.update`.

Reply:
Now that you mention it, I have the same use-case for Blosc, as one might want to manually set the typewidth or bit-shuffle versus byte-shuffle, depending on the individual array.  I agree that a `config` argument to `set_array_compression` is probably the way to go.

Do we need a way to pass options for decompression, like number of threads?  For blosc, most of the time just a global `import blosc; blosc.set_nthreads(n)` would be acceptable, so maybe we don't need a full decompression options mechanism.

Reply:
That's a good question, but yeah, maybe we defer decompression config until someone needs it?

Reply:
I think that makes sense.  I'll put together a version that removes the `AsdfConfig` additions and uses `set_array_compression` args instead.